In [1]:
import pandas as pd
import numpy as np

In [2]:
df_xtrain = pd.read_csv("Training/X_train.csv")
df_xtrain.shape

(33050, 44)

In [3]:
df_xtrain.head()

Unique_ID  C1  C2  C3  C4  C5     C6  C7    C8     N1  ...  N26  N27  \
0   Candidate_5926   1   0  11  31   0  False   0  True  23.75  ...  NaN  NaN   
1  Candidate_48134   1   4   2  66   2  False   1  True  11.05  ...  NaN  NaN   
2  Candidate_51717   1   0  19   2   0  False   0  True  29.00  ...  NaN  NaN   
3  Candidate_26401   1   1  16  47   1  False   4  True  17.99  ...  NaN  NaN   
4  Candidate_34872   1   1  13   1   1   True   6  True  27.50  ...  NaN  NaN   

   N28  N29  N30  N31  N32    N33     N34   N35  
0  NaN  NaN  NaN  NaN  NaN   58.0  113.39  12.0  
1  NaN  NaN  NaN  NaN  NaN  160.0  262.10  17.0  
2  NaN  NaN  NaN  NaN  NaN   24.0   50.29  18.0  
3  NaN  NaN  NaN  NaN  NaN   70.0  126.52  27.0  
4  NaN  NaN  NaN  NaN  NaN  100.0  205.47  21.0  

[5 rows x 44 columns]

In [4]:
df_xtrain.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33050 entries, 0 to 33049
Data columns (total 44 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Unique_ID  33050 non-null  object 
 1   C1         33050 non-null  int64  
 2   C2         33050 non-null  int64  
 3   C3         33050 non-null  int64  
 4   C4         33050 non-null  int64  
 5   C5         33050 non-null  int64  
 6   C6         33050 non-null  bool   
 7   C7         33050 non-null  int64  
 8   C8         33050 non-null  bool   
 9   N1         33050 non-null  float64
 10  N2         28438 non-null  float64
 11  N3         32685 non-null  float64
 12  N4         28444 non-null  float64
 13  N5         28444 non-null  float64
 14  N6         32685 non-null  float64
 15  N7         32623 non-null  float64
 16  N8         33050 non-null  int64  
 17  N9         33050 non-null  int64  
 18  N10        32623 non-null  float64
 19  N10.1      32623 non-null  float64
 20  N11   

In [5]:
df_xtrain.isnull().sum()

Unique_ID        0
C1               0
C2               0
C3               0
C4               0
C5               0
C6               0
C7               0
C8               0
N1               0
N2            4612
N3             365
N4            4606
N5            4606
N6             365
N7             427
N8               0
N9               0
N10            427
N10.1          427
N11            716
N12           4614
N14            608
N15            427
N16           4606
N17           4606
N18           4606
N19           4571
N20           4571
N21           4571
N22           4571
N23           2486
N24              0
N25          26779
N26          26779
N27          26779
N28          26779
N29          26779
N30          26779
N31          26779
N32          26830
N33              0
N34              0
N35            427
dtype: int64

In [6]:
df_xtrain.describe()

C1            C2            C3            C4            C5  \
count  33050.000000  33050.000000  33050.000000  33050.000000  33050.000000   
mean       1.119818      2.403873     12.856097     14.437398      1.719395   
std        0.399382      3.399024     12.013457     15.527474      1.320860   
min        1.000000      0.000000      0.000000      0.000000      0.000000   
25%        1.000000      0.000000      2.000000      1.000000      1.000000   
50%        1.000000      1.000000     10.000000     10.000000      2.000000   
75%        1.000000      3.000000     20.000000     26.000000      2.000000   
max        3.000000     20.000000     51.000000     67.000000      7.000000   

                 C7           N1            N2            N3            N4  \
count  33050.000000  33050.00000  28438.000000  32685.000000  28444.000000   
mean       3.124932     19.97653     81.342710      3.359361      9.574075   
std        2.158905      8.11087     85.397795      0.411040      5.584756   
min        0.000000      0.00000      0.000000      0.000000      0.000000   
25%        1.000000     13.45000     21.000000      3.200000      6.000000   
50%        4.000000     19.27500     53.000000      3.400000      9.000000   
75%        4.000000     26.90000    112.000000      3.600000     13.000000   
max        7.000000     48.00000    755.000000      4.400000     59.000000   

       ...          N26          N27          N28          N29           N30  \
count  ...  6271.000000  6271.000000  6271.000000  6271.000000   6271.000000   
mean   ...    20.084197    19.512199     0.526391     0.045288   7192.813526   
std    ...    16.045120    15.637882     2.144069     0.443838   6460.139952   
min    ...     0.000000     0.000000     0.000000     0.000000   1000.000000   
25%    ...     9.000000     9.000000     0.000000     0.000000   3000.000000   
50%    ...    15.000000    14.000000     0.000000     0.000000   5000.000000   
75%    ...    28.000000    27.000000     0.000000     0.000000   9500.000000   
max    ...   120.000000   111.000000    42.000000    11.000000  60001.000000   

                N31          N32           N33           N34           N35  
count   6271.000000  6220.000000  33050.000000  33050.000000  32623.000000  
mean    2379.621620     0.744373    125.629394    224.313536     19.587254  
std     3379.065104    50.764684    104.391919    186.513844      7.352667  
min        0.000000  -194.000000     20.000000      0.000000      2.000000  
25%        0.000000   -31.000000     54.000000     98.740000     14.000000  
50%     1108.440000     0.000000     90.000000    172.600000     19.000000  
75%     3408.750000    32.000000    160.000000    300.947500     24.000000  
max    22586.670000   257.000000    700.000000   2153.380000     64.000000  

[8 rows x 41 columns]

In [7]:
df_test = pd.read_csv("Test/X_test.csv")
df_test.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11017 entries, 0 to 11016
Data columns (total 44 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Unique_ID  11017 non-null  object 
 1   C1         11017 non-null  int64  
 2   C2         11017 non-null  int64  
 3   C3         11017 non-null  int64  
 4   C4         11017 non-null  int64  
 5   C5         11017 non-null  int64  
 6   C6         11017 non-null  bool   
 7   C7         11017 non-null  int64  
 8   C8         11017 non-null  bool   
 9   N1         11017 non-null  float64
 10  N2         9528 non-null   float64
 11  N3         10905 non-null  float64
 12  N4         9529 non-null   float64
 13  N5         9529 non-null   float64
 14  N6         10905 non-null  float64
 15  N7         10888 non-null  float64
 16  N8         11017 non-null  int64  
 17  N9         11017 non-null  int64  
 18  N10        10888 non-null  float64
 19  N10.1      10888 non-null  float64
 20  N11   

In [8]:
(df_test.isnull().sum()/len(df_test)*100).sort_values(ascending = False).head(50)

N32          80.829627
N25          80.593628
N31          80.593628
N30          80.593628
N29          80.593628
N28          80.593628
N27          80.593628
N26          80.593628
N12          13.533630
N2           13.515476
N4           13.506399
N5           13.506399
N16          13.506399
N17          13.506399
N18          13.506399
N22          13.352092
N21          13.352092
N20          13.352092
N19          13.352092
N23           7.706272
N11           1.978760
N14           1.706454
N15           1.170918
N35           1.170918
N10           1.170918
N7            1.170918
N10.1         1.170918
N6            1.016611
N3            1.016611
C1            0.000000
C5            0.000000
N34           0.000000
N33           0.000000
C2            0.000000
C3            0.000000
C4            0.000000
C8            0.000000
C6            0.000000
C7            0.000000
N1            0.000000
N24           0.000000
N8            0.000000
N9            0.000000
Unique_ID  

In [9]:
(df_xtrain.isnull().sum()/len(df_xtrain)*100).sort_values(ascending = False).head(50)

N32          81.180030
N25          81.025719
N31          81.025719
N30          81.025719
N29          81.025719
N28          81.025719
N27          81.025719
N26          81.025719
N12          13.960666
N2           13.954614
N4           13.936460
N5           13.936460
N16          13.936460
N17          13.936460
N18          13.936460
N22          13.830560
N21          13.830560
N20          13.830560
N19          13.830560
N23           7.521936
N11           2.166415
N14           1.839637
N15           1.291982
N35           1.291982
N10           1.291982
N7            1.291982
N10.1         1.291982
N6            1.104387
N3            1.104387
C1            0.000000
C5            0.000000
N34           0.000000
N33           0.000000
C2            0.000000
C3            0.000000
C4            0.000000
C8            0.000000
C6            0.000000
C7            0.000000
N1            0.000000
N24           0.000000
N8            0.000000
N9            0.000000
Unique_ID  

In [10]:
null_col = df_xtrain.isnull().sum().sort_values(ascending = False)
null_col = null_col[null_col.values >(0.35*len(df_xtrain))]
null_col.shape

(8,)

There are 8 volumns with null cells more than 35%
Therefore we will drop these columns and focus on other parts

In [11]:
label = list(null_col.index.values) #Making list of column names having null values greater than 35%
df_xtrain.drop(labels = label,axis=1,inplace = True) #Droping those columns
df_xtrain.shape

(33050, 36)

In [12]:
null = (df_xtrain.isnull().sum()/len(df_xtrain)*100).sort_values(ascending = True)
null.head(20)

Unique_ID    0.000000
N33          0.000000
N24          0.000000
N34          0.000000
N8           0.000000
N1           0.000000
C8           0.000000
N9           0.000000
C6           0.000000
C5           0.000000
C1           0.000000
C2           0.000000
C7           0.000000
C4           0.000000
C3           0.000000
N3           1.104387
N6           1.104387
N15          1.291982
N10.1        1.291982
N35          1.291982
dtype: float64

In [13]:
df_ytrain = pd.read_csv("Training/y_train.csv")
df_ytrain.head()

Unique_ID  Dependent_Variable
0   Candidate_5926                   1
1  Candidate_48134                   0
2  Candidate_51717                   1
3  Candidate_26401                   0
4  Candidate_34872                   0

In [14]:
print(df_ytrain.groupby(['Dependent_Variable'])['Dependent_Variable'].count())

Dependent_Variable
0    22844
1    10206
Name: Dependent_Variable, dtype: int64


As we can see, 22k+ rpws have 0 as the dependent variable(our target) and 10k+ have 1 as our dependent variable 

In [15]:
df_xtrain["C6"] = df_xtrain["C6"].astype(int)
df_xtrain["C8"] = df_xtrain["C8"].astype(int)

In [16]:
df_ytrain.info()
df_xtrain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33050 entries, 0 to 33049
Data columns (total 2 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Unique_ID           33050 non-null  object
 1   Dependent_Variable  33050 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 516.5+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33050 entries, 0 to 33049
Data columns (total 36 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Unique_ID  33050 non-null  object 
 1   C1         33050 non-null  int64  
 2   C2         33050 non-null  int64  
 3   C3         33050 non-null  int64  
 4   C4         33050 non-null  int64  
 5   C5         33050 non-null  int64  
 6   C6         33050 non-null  int64  
 7   C7         33050 non-null  int64  
 8   C8         33050 non-null  int64  
 9   N1         33050 non-null  float64
 10  N2         28438 non-null  float64
 11  N3         326

As we can look into the training data and the test data, many of the columns have huge percentage of null values
after taking out the ones with more than 35% nulls, we have 3 choices here onwards
<br>
dataframe1: only keep those columns with 100% filled columns and ignore alll the rest (this gave a roc_auc score of 0.64)
<br>
dataframe2: keep all the columns and remove the rows that have a null values 
<br>
dataframe3: keep all the rows and columns, fill null cells with either 0 (as them being blank shows some kind of small significance) or the mean of the column if they are significant

<H2>The following is dataframe 1 </H2>

We do a logistic regression on the data and use only those columnsthat have 100% values, for that 
we have to clean the dataset off the null values

In [17]:
hasnull_col = df_xtrain.isnull().sum().sort_values(ascending = False)
hasnull_col = hasnull_col[hasnull_col.values >0]
hasnull_col.shape

(21,)

In [18]:
label = list(hasnull_col.index.values) #Making list of column names having null values
df1_xtrain = df_xtrain.drop(labels = label,axis=1) #Droping those columns
df1_xtrain.shape
df_xtrain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33050 entries, 0 to 33049
Data columns (total 36 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Unique_ID  33050 non-null  object 
 1   C1         33050 non-null  int64  
 2   C2         33050 non-null  int64  
 3   C3         33050 non-null  int64  
 4   C4         33050 non-null  int64  
 5   C5         33050 non-null  int64  
 6   C6         33050 non-null  int64  
 7   C7         33050 non-null  int64  
 8   C8         33050 non-null  int64  
 9   N1         33050 non-null  float64
 10  N2         28438 non-null  float64
 11  N3         32685 non-null  float64
 12  N4         28444 non-null  float64
 13  N5         28444 non-null  float64
 14  N6         32685 non-null  float64
 15  N7         32623 non-null  float64
 16  N8         33050 non-null  int64  
 17  N9         33050 non-null  int64  
 18  N10        32623 non-null  float64
 19  N10.1      32623 non-null  float64
 20  N11   

In [19]:
df1_xtrain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33050 entries, 0 to 33049
Data columns (total 15 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Unique_ID  33050 non-null  object 
 1   C1         33050 non-null  int64  
 2   C2         33050 non-null  int64  
 3   C3         33050 non-null  int64  
 4   C4         33050 non-null  int64  
 5   C5         33050 non-null  int64  
 6   C6         33050 non-null  int64  
 7   C7         33050 non-null  int64  
 8   C8         33050 non-null  int64  
 9   N1         33050 non-null  float64
 10  N8         33050 non-null  int64  
 11  N9         33050 non-null  int64  
 12  N24        33050 non-null  float64
 13  N33        33050 non-null  float64
 14  N34        33050 non-null  float64
dtypes: float64(4), int64(10), object(1)
memory usage: 3.8+ MB


In [20]:
df1_train = pd.merge(df1_xtrain,df_ytrain, on='Unique_ID')
df1_train.head()

Unique_ID  C1  C2  C3  C4  C5  C6  C7  C8     N1  N8    N9  \
0   Candidate_5926   1   0  11  31   0   0   0   1  23.75   0     0   
1  Candidate_48134   1   4   2  66   2   0   1   1  11.05   4    72   
2  Candidate_51717   1   0  19   2   0   0   0   1  29.00   2    15   
3  Candidate_26401   1   1  16  47   1   0   4   1  17.99   5   190   
4  Candidate_34872   1   1  13   1   1   1   6   1  27.50  10  1504   

            N24    N33     N34  Dependent_Variable  
0   1750.000000   58.0  113.39                   1  
1  10833.333330  160.0  262.10                   0  
2   6250.000000   24.0   50.29                   1  
3   2413.666667   70.0  126.52                   0  
4   7666.666667  100.0  205.47                   0

In [21]:
df1_train.describe()

C1            C2            C3            C4            C5  \
count  33050.000000  33050.000000  33050.000000  33050.000000  33050.000000   
mean       1.119818      2.403873     12.856097     14.437398      1.719395   
std        0.399382      3.399024     12.013457     15.527474      1.320860   
min        1.000000      0.000000      0.000000      0.000000      0.000000   
25%        1.000000      0.000000      2.000000      1.000000      1.000000   
50%        1.000000      1.000000     10.000000     10.000000      2.000000   
75%        1.000000      3.000000     20.000000     26.000000      2.000000   
max        3.000000     20.000000     51.000000     67.000000      7.000000   

                 C6            C7            C8           N1            N8  \
count  33050.000000  33050.000000  33050.000000  33050.00000  33050.000000   
mean       0.467958      3.124932      0.923328     19.97653      6.082723   
std        0.498980      2.158905      0.266074      8.11087      4.535870   
min        0.000000      0.000000      0.000000      0.00000      0.000000   
25%        0.000000      1.000000      1.000000     13.45000      3.000000   
50%        0.000000      4.000000      1.000000     19.27500      5.000000   
75%        1.000000      4.000000      1.000000     26.90000      8.000000   
max        1.000000      7.000000      1.000000     48.00000     51.000000   

                 N9            N24           N33           N34  \
count  33050.000000   33050.000000  33050.000000  33050.000000   
mean     325.585870    5050.306727    125.629394    224.313536   
std      452.149358    6237.726330    104.391919    186.513844   
min        0.000000       0.000000     20.000000      0.000000   
25%       60.000000    2833.333333     54.000000     98.740000   
50%      183.000000    4166.666667     90.000000    172.600000   
75%      420.000000    6250.000000    160.000000    300.947500   
max    14985.000000  618547.833300    700.000000   2153.380000   

       Dependent_Variable  
count        33050.000000  
mean             0.308805  
std              0.462007  
min              0.000000  
25%              0.000000  
50%              0.000000  
75%              1.000000  
max              1.000000

Slice the data into test and train sets

In [22]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df1_xtrain.drop(['Unique_ID'], axis = 1), df_ytrain['Dependent_Variable'] , test_size=0.25, random_state=0)

In [23]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(solver='lbfgs', max_iter=3000)
model.fit(X_train, y_train)

LogisticRegression(max_iter=3000)

In [24]:
predictions = model.predict(X_test)

In [25]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, predictions)

TN, FP, FN, TP = confusion_matrix(y_test, predictions).ravel()

print('True Positive(TP)  = ', TP)
print('False Positive(FP) = ', FP)
print('True Negative(TN)  = ', TN)
print('False Negative(FN) = ', FN)

accuracy =  (TP+TN) /(TP+FP+TN+FN)

print('Accuracy of the binary classification = {:0.3f}'.format(accuracy))

True Positive(TP)  =  572
False Positive(FP) =  463
True Negative(TN)  =  5166
False Negative(FN) =  2062
Accuracy of the binary classification = 0.694


In [26]:
from sklearn.metrics import roc_auc_score, auc, roc_curve
fpr, tpr, thresholds = roc_curve(y_test, predictions, pos_label=1)
auc_algo =  auc(fpr, tpr)
print(auc_algo)

0.5674537961227739


Trying out all the different binary classifiers except nn because they scare me

In [27]:
def models(X_train, X_test, y_train, y_test):
    models = {}

    # Logistic Regression
    from sklearn.linear_model import LogisticRegression
    models['Logistic Regression'] = LogisticRegression(solver='lbfgs', max_iter=3000)

    # Support Vector Machines
    from sklearn.svm import LinearSVC
    models['Support Vector Machines'] = LinearSVC()

    # Decision Trees
    from sklearn.tree import DecisionTreeClassifier
    models['Decision Trees'] = DecisionTreeClassifier()

    # Random Forest
    from sklearn.ensemble import RandomForestClassifier
    models['Random Forest'] = RandomForestClassifier()

    # Naive Bayes
    from sklearn.naive_bayes import GaussianNB
    models['Naive Bayes'] = GaussianNB()

    # K-Nearest Neighbors
    from sklearn.neighbors import KNeighborsClassifier
    models['K-Nearest Neighbor'] = KNeighborsClassifier()
    from sklearn.metrics import accuracy_score, precision_score, recall_score

    accuracy, precision, recall,aucAlgo = {}, {}, {}, {}

    for key in models.keys():
        
        # Fit the classifier model
        models[key].fit(X_train, y_train)
        
        # Prediction 
        predictions = models[key].predict(X_test)
        
        # Calculate Accuracy, Precision and Recall Metrics
        accuracy[key] = accuracy_score(predictions, y_test)
        precision[key] = precision_score(predictions, y_test)
        recall[key] = recall_score(predictions, y_test)
        
    
    import pandas as pd

    df_model = pd.DataFrame(index=models.keys(), columns=['Accuracy', 'Precision', 'Recall','RocAuc'])
    df_model['Accuracy'] = accuracy.values()
    df_model['Precision'] = precision.values()
    df_model['Recall'] = recall.values()

    return df_model



In [28]:
models(X_train, X_test, y_train, y_test)

/Users/namanjoshi/opt/anaconda3/lib/python3.9/site-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Accuracy  Precision    Recall RocAuc
Logistic Regression      0.694421   0.217160  0.552657    NaN
Support Vector Machines  0.684013   0.029992  0.585185    NaN
Decision Trees           0.639840   0.429385  0.434332    NaN
Random Forest            0.725523   0.331055  0.632801    NaN
Naive Bayes              0.638146   0.632118  0.451709    NaN
K-Nearest Neighbor       0.648191   0.273349  0.420315    NaN

Trying out NN because others dont work so good


# Neural Networks
from keras.models import Sequential
from keras.layers import Dense

model = Sequential() 
model.add(Dense(128, activation='relu', input_dim=30))
model.add(Dense(1, activation='sigmoid')) 
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy']) 
model.summary()

hist = model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=100)

Coded the NN on Google collab as due to macos some problem sustained and nn didnt work

Results from collab on NN:
Epoch 1/10<br>
248/248 [==============================] - 3s 8ms/step - loss: 11.0341 - accuracy: 0.5929 - val_loss: 7.7712 - val_accuracy: 0.6820<br>
Epoch 2/10<br>
248/248 [==============================] - 1s 5ms/step - loss: 8.6967 - accuracy: 0.5979 - val_loss: 10.3662 - val_accuracy: 0.4625<br>
Epoch 3/10<br>
248/248 [==============================] - 1s 2ms/step - loss: 6.9906 - accuracy: 0.6044 - val_loss: 5.2547 - val_accuracy: 0.6702<br>
Epoch 4/10<br>
248/248 [==============================] - 1s 2ms/step - loss: 10.2708 - accuracy: 0.6041 - val_loss: 7.1569 - val_accuracy: 0.3840<br>
Epoch 5/10<br>
248/248 [==============================] - 1s 2ms/step - loss: 8.0592 - accuracy: 0.6034 - val_loss: 12.8009 - val_accuracy: 0.6828<br>
Epoch 6/10<br>
248/248 [==============================] - 1s 4ms/step - loss: 5.9560 - accuracy: 0.6175 - val_loss: 11.8021 - val_accuracy: 0.6835<br>
Epoch 7/10<br>
248/248 [==============================] - 1s 4ms/step - loss: 6.9240 - accuracy: 0.6119 - val_loss: 9.1898 - val_accuracy: 0.6846<br>
Epoch 8/10<br>
248/248 [==============================] - 1s 4ms/step - loss: 6.2201 - accuracy: 0.6095 - val_loss: 5.8085 - val_accuracy: 0.6789<br>
Epoch 9/10<br>
248/248 [==============================] - 1s 5ms/step - loss: 7.3723 - accuracy: 0.6114 - val_loss: 3.6043 - val_accuracy: 0.6772<br>
Epoch 10/10<br>
248/248 [==============================] - 1s 4ms/step - loss: 8.5643 - accuracy: 0.6094 - val_loss: 3.4606 - val_accuracy: 0.5227<br>


<h2>The following is for dataframe2</h2>

In [29]:
df2_train = pd.merge(df_xtrain,df_ytrain, on='Unique_ID')
df2_train.head()

Unique_ID  C1  C2  C3  C4  C5  C6  C7  C8     N1  ...      N19   N20  \
0   Candidate_5926   1   0  11  31   0   0   0   1  23.75  ...      NaN   NaN   
1  Candidate_48134   1   4   2  66   2   0   1   1  11.05  ...  25856.0  17.0   
2  Candidate_51717   1   0  19   2   0   0   0   1  29.00  ...      NaN   NaN   
3  Candidate_26401   1   1  16  47   1   0   4   1  17.99  ...   1006.0   6.0   
4  Candidate_34872   1   1  13   1   1   1   6   1  27.50  ...   3398.0  31.0   

    N21  N22     N23           N24    N33     N34   N35  Dependent_Variable  
0   NaN  NaN  27.816   1750.000000   58.0  113.39  12.0                   1  
1  0.88  1.0  40.000  10833.333330  160.0  262.10  17.0                   0  
2   NaN  NaN  20.000   6250.000000   24.0   50.29  18.0                   1  
3  1.00  0.0  26.000   2413.666667   70.0  126.52  27.0                   0  
4  0.96  0.0  44.000   7666.666667  100.0  205.47  21.0                   0  

[5 rows x 37 columns]

In [30]:
df2_train.shape

(33050, 37)

In [31]:
print(df2_train.groupby(['Dependent_Variable'])['Dependent_Variable'].count())

Dependent_Variable
0    22844
1    10206
Name: Dependent_Variable, dtype: int64


In [32]:
df2_train.dropna(inplace=True)
df2_train.shape

(26027, 37)

In [33]:
print(df2_train.groupby(['Dependent_Variable'])['Dependent_Variable'].count()/len(df2_train))

Dependent_Variable
0    0.712184
1    0.287816
Name: Dependent_Variable, dtype: float64


An observation here is that before the ration of dependent variable was 1:2 and now it is 2:5 which may cause the models to have bias

In [34]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df2_train.drop(['Unique_ID','Dependent_Variable'], axis = 1), df2_train['Dependent_Variable'] , test_size=0.25, random_state=0)

In [35]:
models(X_train, X_test, y_train, y_test)

/Users/namanjoshi/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Users/namanjoshi/opt/anaconda3/lib/python3.9/site-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Accuracy  Precision    Recall RocAuc
Logistic Regression      0.729368   0.174858  0.670635    NaN
Support Vector Machines  0.710466   0.041386  0.720721    NaN
Decision Trees           0.660058   0.427315  0.427758    NaN
Random Forest            0.744583   0.262287  0.682369    NaN
Naive Bayes              0.714308   0.270047  0.538144    NaN
K-Nearest Neighbor       0.671738   0.227108  0.406105    NaN

<h5>All 4 metrics have diminished in scores and the precision is gone too.</h5>

<h2>We now try with dataframe3</h2>
<p>This the one where we fill all na values with 0<p>

In [36]:
df3_train = pd.merge(df_xtrain,df_ytrain, on='Unique_ID')
df3_train.head()

Unique_ID  C1  C2  C3  C4  C5  C6  C7  C8     N1  ...      N19   N20  \
0   Candidate_5926   1   0  11  31   0   0   0   1  23.75  ...      NaN   NaN   
1  Candidate_48134   1   4   2  66   2   0   1   1  11.05  ...  25856.0  17.0   
2  Candidate_51717   1   0  19   2   0   0   0   1  29.00  ...      NaN   NaN   
3  Candidate_26401   1   1  16  47   1   0   4   1  17.99  ...   1006.0   6.0   
4  Candidate_34872   1   1  13   1   1   1   6   1  27.50  ...   3398.0  31.0   

    N21  N22     N23           N24    N33     N34   N35  Dependent_Variable  
0   NaN  NaN  27.816   1750.000000   58.0  113.39  12.0                   1  
1  0.88  1.0  40.000  10833.333330  160.0  262.10  17.0                   0  
2   NaN  NaN  20.000   6250.000000   24.0   50.29  18.0                   1  
3  1.00  0.0  26.000   2413.666667   70.0  126.52  27.0                   0  
4  0.96  0.0  44.000   7666.666667  100.0  205.47  21.0                   0  

[5 rows x 37 columns]

In [37]:
df3_train.shape

(33050, 37)

In [38]:
dftry= df3_train.replace(np.nan,0)

In [39]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(dftry.drop(['Unique_ID','Dependent_Variable'], axis = 1), dftry['Dependent_Variable'] , test_size=0.25, random_state=0)

In [40]:
models(X_train, X_test, y_train, y_test)

/Users/namanjoshi/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Users/namanjoshi/opt/anaconda3/lib/python3.9/site-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Accuracy  Precision    Recall RocAuc
Logistic Regression      0.713905   0.260440  0.622505    NaN
Support Vector Machines  0.375772   0.779423  0.309653    NaN
Decision Trees           0.643471   0.428626  0.439300    NaN
Random Forest            0.730848   0.320425  0.660407    NaN
Naive Bayes              0.699746   0.314351  0.550898    NaN
K-Nearest Neighbor       0.654605   0.252088  0.428941    NaN

Logistic Regression	0.683771	0.037585	0.559322	0.511864
Support Vector Machines	0.395740	0.690585	0.303318	0.474178
Decision Trees	0.641656	0.429385	0.436848	0.585185
Random Forest	0.732785	0.319666	0.669316	0.622881
Naive Bayes	0.699746	0.314351	0.550898	0.597218
K-Nearest Neighbor	0.654605	0.252088	0.428941	0.547522

<h2>creating df4 just to check with mean values in place of nan</h2>

In [41]:
df4_train = pd.merge(df_xtrain,df_ytrain, on='Unique_ID')
df4_train.head()

Unique_ID  C1  C2  C3  C4  C5  C6  C7  C8     N1  ...      N19   N20  \
0   Candidate_5926   1   0  11  31   0   0   0   1  23.75  ...      NaN   NaN   
1  Candidate_48134   1   4   2  66   2   0   1   1  11.05  ...  25856.0  17.0   
2  Candidate_51717   1   0  19   2   0   0   0   1  29.00  ...      NaN   NaN   
3  Candidate_26401   1   1  16  47   1   0   4   1  17.99  ...   1006.0   6.0   
4  Candidate_34872   1   1  13   1   1   1   6   1  27.50  ...   3398.0  31.0   

    N21  N22     N23           N24    N33     N34   N35  Dependent_Variable  
0   NaN  NaN  27.816   1750.000000   58.0  113.39  12.0                   1  
1  0.88  1.0  40.000  10833.333330  160.0  262.10  17.0                   0  
2   NaN  NaN  20.000   6250.000000   24.0   50.29  18.0                   1  
3  1.00  0.0  26.000   2413.666667   70.0  126.52  27.0                   0  
4  0.96  0.0  44.000   7666.666667  100.0  205.47  21.0                   0  

[5 rows x 37 columns]

In [42]:
hasnull_col = df4_train.isnull().sum().sort_values(ascending = False)
hasnull_col = hasnull_col[hasnull_col.values >0]
hasnull_col.shape

(21,)

In [43]:
hasnull_col.index

Index(['N12', 'N2', 'N4', 'N17', 'N16', 'N5', 'N18', 'N19', 'N22', 'N21',
       'N20', 'N23', 'N11', 'N14', 'N10.1', 'N10', 'N7', 'N35', 'N15', 'N6',
       'N3'],
      dtype='object')

In [44]:
df4_train[hasnull_col.index] = df4_train[hasnull_col.index].fillna(value=df4_train[hasnull_col.index].mean())

df4_train.head(15)

Unique_ID  C1  C2  C3  C4  C5  C6  C7  C8     N1  ...           N19  \
0    Candidate_5926   1   0  11  31   0   0   0   1  23.75  ...  11037.365322   
1   Candidate_48134   1   4   2  66   2   0   1   1  11.05  ...  25856.000000   
2   Candidate_51717   1   0  19   2   0   0   0   1  29.00  ...  11037.365322   
3   Candidate_26401   1   1  16  47   1   0   4   1  17.99  ...   1006.000000   
4   Candidate_34872   1   1  13   1   1   1   6   1  27.50  ...   3398.000000   
5   Candidate_29660   1   0  30  13   2   0   2   1  13.50  ...   4110.000000   
6   Candidate_29517   1   0  41   0   2   1   2   1   8.50  ...  38757.000000   
7   Candidate_34848   1   1  34  17   1   0   7   1  16.20  ...    791.000000   
8   Candidate_12081   1   0  35  31   0   0   0   1  11.75  ...  11037.365322   
9   Candidate_34016   1   4  17   0   2   1   4   1  10.50  ...   8019.000000   
10  Candidate_52651   2   1   2   1   1   0   4   1  28.27  ...  15400.000000   
11   Candidate_8149   1   1   7   1   1   0   4   0  30.73  ...   2077.000000   
12  Candidate_22621   1   4   5   0   2   1   6   1   7.00  ...  30937.000000   
13  Candidate_36516   1   1  13  35   1   1   2   1  15.99  ...   9060.000000   
14  Candidate_20236   1   0   6  10   2   0   7   1  14.25  ...    163.000000   

          N20       N21       N22        N23           N24     N33     N34  \
0   21.764142  0.857199  0.906422  27.816000   1750.000000   58.00  113.39   
1   17.000000  0.880000  1.000000  40.000000  10833.333330  160.00  262.10   
2   21.764142  0.857199  0.906422  20.000000   6250.000000   24.00   50.29   
3    6.000000  1.000000  0.000000  26.000000   2413.666667   70.00  126.52   
4   31.000000  0.960000  0.000000  44.000000   7666.666667  100.00  205.47   
5   11.000000  0.900000  2.000000  48.000000   4250.000000  150.00  254.51   
6   23.000000  1.000000  2.000000  30.000000   4875.000000   51.00   80.50   
7    4.000000  1.000000  1.000000  42.000000   1960.000000   53.00   93.43   
8   21.764142  0.857199  0.906422  32.000000   3750.000000   36.00    0.00   
9   14.000000  1.000000  0.000000  30.000000   3000.000000   20.00   32.50   
10  17.000000  1.000000  0.000000  90.000000   3500.000000  140.00  219.09   
11  22.000000  0.810000  2.000000  58.433363   3889.583333   80.00  171.41   
12  27.000000  0.920000  0.000000  20.000000   7916.666667   60.02   92.66   
13   8.000000  1.000000  0.000000  28.000000   4583.333333  100.00  175.76   
14   5.000000  0.400000  0.000000   4.000000   2080.000000   30.00   51.45   

     N35  Dependent_Variable  
0   12.0                   1  
1   17.0                   0  
2   18.0                   1  
3   27.0                   0  
4   21.0                   0  
5   13.0                   0  
6   19.0                   0  
7    3.0                   0  
8   18.0                   0  
9   14.0                   0  
10  11.0                   1  
11  20.0                   1  
12  26.0                   0  
13  14.0                   0  
14  15.0                   0  

[15 rows x 37 columns]

In [45]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df4_train.drop(['Unique_ID','Dependent_Variable'], axis = 1), df4_train['Dependent_Variable'] , test_size=0.25, random_state=0)

In [46]:
models(X_train, X_test, y_train, y_test)

/Users/namanjoshi/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Users/namanjoshi/opt/anaconda3/lib/python3.9/site-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Accuracy  Precision    Recall RocAuc
Logistic Regression      0.717173   0.291572  0.619855    NaN
Support Vector Machines  0.547622   0.498861  0.352090    NaN
Decision Trees           0.642503   0.442673  0.439668    NaN
Random Forest            0.731453   0.323083  0.661228    NaN
Naive Bayes              0.697688   0.296507  0.547686    NaN
K-Nearest Neighbor       0.655089   0.250569  0.429688    NaN

<h4>Across all analysis and prediction, the best accuracy score is given by RandomForrest and then by Naive bayes followed by Logistic Regression</h4>

<h5>We will therefore move ahead with randomforrest and with different parameters test for better scores</h5>

<h3>Random Forrest </h3>
<h4>Hyperparameter Tuning- RandomizedSearchCV</h4>
<h5>Using the dataset df1</h5>

In [47]:
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV

In [48]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df1_xtrain.drop(['Unique_ID'], axis = 1), df_ytrain['Dependent_Variable'] , test_size=0.25, random_state=22)

In [49]:
param_grid = {
    'n_estimators': [25, 50, 100, 150],
    'max_features': ['sqrt', 'log2', None],
    'max_depth': [3, 6, 9],
    'max_leaf_nodes': [3, 6, 9],
}

In [50]:
random_search = RandomizedSearchCV(RandomForestClassifier(),param_grid,cv=3)
random_search.fit(X_train, y_train)
print(random_search.best_estimator_)

RandomForestClassifier(max_depth=3, max_features=None, max_leaf_nodes=9,
                       n_estimators=50)


In [51]:
X_train.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 24787 entries, 12250 to 29484
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   C1      24787 non-null  int64  
 1   C2      24787 non-null  int64  
 2   C3      24787 non-null  int64  
 3   C4      24787 non-null  int64  
 4   C5      24787 non-null  int64  
 5   C6      24787 non-null  int64  
 6   C7      24787 non-null  int64  
 7   C8      24787 non-null  int64  
 8   N1      24787 non-null  float64
 9   N8      24787 non-null  int64  
 10  N9      24787 non-null  int64  
 11  N24     24787 non-null  float64
 12  N33     24787 non-null  float64
 13  N34     24787 non-null  float64
dtypes: float64(4), int64(10)
memory usage: 2.8 MB


In [52]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8263 entries, 2637 to 4418
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   C1      8263 non-null   int64  
 1   C2      8263 non-null   int64  
 2   C3      8263 non-null   int64  
 3   C4      8263 non-null   int64  
 4   C5      8263 non-null   int64  
 5   C6      8263 non-null   int64  
 6   C7      8263 non-null   int64  
 7   C8      8263 non-null   int64  
 8   N1      8263 non-null   float64
 9   N8      8263 non-null   int64  
 10  N9      8263 non-null   int64  
 11  N24     8263 non-null   float64
 12  N33     8263 non-null   float64
 13  N34     8263 non-null   float64
dtypes: float64(4), int64(10)
memory usage: 968.3 KB


In [53]:
model_grid = random_search.best_estimator_
model_grid.fit(X_train, y_train)
y_pred_grid = model.predict(X_test)
#print(classification_report(y_pred_grid, y_test))

In [54]:
print(classification_report(y_pred_grid, y_test))

              precision    recall  f1-score   support

           0       0.91      0.72      0.81      7271
           1       0.19      0.49      0.28       992

    accuracy                           0.69      8263
   macro avg       0.55      0.61      0.54      8263
weighted avg       0.83      0.69      0.74      8263



Considering after all the different variations, the best fit model (looking at variation of data and scores throughout) is random forrest without any hyper tuning, we will use it as our final model

the dataset it will be trained on is the df1 that has all the 100% values and the text dataset will be similarly changed to the same dimension.

In [55]:
df_test = pd.read_csv("Test/X_test.csv")
df_test.shape

(11017, 44)

In [56]:
X_test.shape

(8263, 14)

In [57]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11017 entries, 0 to 11016
Data columns (total 44 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Unique_ID  11017 non-null  object 
 1   C1         11017 non-null  int64  
 2   C2         11017 non-null  int64  
 3   C3         11017 non-null  int64  
 4   C4         11017 non-null  int64  
 5   C5         11017 non-null  int64  
 6   C6         11017 non-null  bool   
 7   C7         11017 non-null  int64  
 8   C8         11017 non-null  bool   
 9   N1         11017 non-null  float64
 10  N2         9528 non-null   float64
 11  N3         10905 non-null  float64
 12  N4         9529 non-null   float64
 13  N5         9529 non-null   float64
 14  N6         10905 non-null  float64
 15  N7         10888 non-null  float64
 16  N8         11017 non-null  int64  
 17  N9         11017 non-null  int64  
 18  N10        10888 non-null  float64
 19  N10.1      10888 non-null  float64
 20  N11   

In [58]:
df_test["C6"] = df_test["C6"].astype(int)
df_test["C8"] = df_test["C8"].astype(int)

In [59]:
test_hasnull_col = df_test.isnull().sum().sort_values(ascending = False)
test_hasnull_col = test_hasnull_col[test_hasnull_col.values >0]
test_hasnull_col.shape

(29,)

In [60]:
label = list(test_hasnull_col.index.values) #Making list of column names having null values greater than 35%
df_test.drop(labels = label,axis=1,inplace = True) #Droping those columns
df_test.shape

(11017, 15)

In [61]:
set(X_test.columns) == set(df_test.drop(["Unique_ID"],axis = 1).columns)

True

<H3>Final Model build and predict probabilities:</H3>

In [62]:
from sklearn.ensemble import RandomForestClassifier
modelRFC= RandomForestClassifier()

In [63]:
modelRFC.fit(df1_xtrain.drop(['Unique_ID'], axis = 1), df_ytrain['Dependent_Variable'])

RandomForestClassifier()

In [64]:
df_pred = pd.DataFrame()

In [65]:
dfd = pd.read_csv("sample_submission_file.csv")

In [66]:
dfd.head()

Unique_ID  Class_1_Probability
0  Candidate_17537                  0.5
1  Candidate_21230                  0.5
2  Candidate_41937                  0.5
3  Candidate_19223                  0.5
4  Candidate_34995                  0.5

In [67]:
df_pred["Unique_ID"] = df_test["Unique_ID"]
df_pred.head()

Unique_ID
0   Candidate_1602
1  Candidate_29650
2  Candidate_31061
3   Candidate_5768
4  Candidate_27059

In [68]:
preds = modelRFC.predict_proba(df_test.drop(["Unique_ID"],axis = 1))[:,1]

In [69]:
print(type(preds))

<class 'numpy.ndarray'>


In [70]:
df_pred["Class_1_Probability"] = preds

In [71]:
df_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11017 entries, 0 to 11016
Data columns (total 2 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Unique_ID            11017 non-null  object 
 1   Class_1_Probability  11017 non-null  float64
dtypes: float64(1), object(1)
memory usage: 172.3+ KB


In [72]:
df_pred.head()

Unique_ID  Class_1_Probability
0   Candidate_1602                 0.49
1  Candidate_29650                 0.61
2  Candidate_31061                 0.41
3   Candidate_5768                 0.22
4  Candidate_27059                 0.49

In [73]:
df_pred.to_csv("Test/final_predictions.csv")